rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,506 @@
+
+*   https://github.com/keras-team/keras/issues/5738
+*   https://github.com/keras-team/keras/issues/4962
+*   https://github.com/keras-team/keras/issues/2525
+
+## Design Proposal
+
+We propose to implement the following common attention layers:
+
+*   `Attention`: Basic dot-product attention, a.k.a. Luong-style attention.
+    Follows
+    [tf.contrib.seq2seq.LuongAttention](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/LuongAttention).
+    This attention has two forms.
+    *   The first is standard dot-product attention, as described in: Minh-Thang
+        Luong, Hieu Pham, Christopher D. Manning. "Effective Approaches to
+        Attention-based Neural Machine Translation." EMNLP 2015.
+        https://arxiv.org/abs/1508.04025.
+    *   The second is the scaled form inspired partly by the normalized form of
+        additive (Bahdanau-style) attention. To enable the second form,
+        construct the object with parameter `scale=True`.
+*   `AdditiveAttention`: Additive attention, a.k.a. Bahdanau-style attention.
+    Follows
+    [tf.contrib.seq2seq.BahdanauAttention](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BahdanauAttention).
+    This attention has two forms.
+    *   The first is additive attention, as described in: Dzmitry Bahdanau,
+        Kyunghyun Cho, Yoshua Bengio. "Neural Machine Translation by Jointly
+        Learning to Align and Translate." ICLR 2015.
+        https://arxiv.org/abs/1409.0473.
+    *   The second is the normalized form. This form is inspired by the weight
+        normalization article: Tim Salimans, Diederik P. Kingma. "Weight
+        Normalization: A Simple Reparameterization to Accelerate Training of
+        Deep Neural Networks." https://arxiv.org/abs/1602.07868. To enable the
+        second form, construct the object with parameter `normalize=True`.
+
+## Detailed Design
+
+According to the general definition of attention (see
+https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1184/lectures/lecture11.pdf),
+"Given a set of vector values, and a vector query, attention is a technique to
+compute a weighted sum of the values, dependent on the query."
+
+There are four input tensors:
+
+*   `query` of shape `[batch_size, Tq, dim]`
+*   `value` of shape `[batch_size, Tv, dim]`
+*   `query_mask` (optional) of shape `[batch_size, Tq]`. Boolean tensor,
+    typically calculated from the query length tensor. Used to mask the output
+    tensor. This is similar to the `mask` argument of
+    [tf.keras.backend.rnn](https://www.tensorflow.org/api_docs/python/tf/keras/backend/rnn).
+*   `value_mask` (optional) of shape `[batch_size, Tv]`. Boolean tensor,
+    typically calculated from the value length tensor. It is used to mask
+    `value` elements beyond this length so they do not contribute to the result.
+
+The output is of shape `[batch_size, Tq, dim]`.
+
+Following the pattern of other Keras layers, we pass the list `[query, value]`
```

Comment:
This feels awkward, specially passing a tuple as the mask argument. Why not have query_mask and value_mask arguments separately, for better self-documenting code?

Similarly for query and value; it's nice if users can tell easily which is which.

Reply:
Francois suggested a list, because the same pattern is used in other places, such as tf.keras.layers.Add. See the code in https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/python/keras/layers/merge.py#L205

Reply:
Let's discuss in the design review.